File: Problem/By section/Dynamic Programing/Problem 198/problem_198.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // Update rule
    // dp[i] = max(dp[i - 1], nums[i] + dp[i - 2])
    // Time: O(n), Space: O(1)
    int rob(vector<int>& nums) {
        if(nums.size() == 1) return nums[0];

        int size = nums.size();
        int prev_2 = nums[0];
        int prev_1 = max(prev_2, nums[1]);
        int result = prev_1;

        for(int i = 2; i < size; i++) {
            result = max(prev_1, prev_2 + nums[i]);
            prev_2 = prev_1;
            prev_1 = result;
        }

        return result;
    }
};

int main(int argc, char** argv) {
    Solution s;
    
    return 0;
}

File: Problem/By section/Dynamic Programing/Problem 198/problem_198.java
import java.util.*;
class Solution {
    // update rule
    // on a certain house
    // you can choose to rob it and skip the one before it
    // or you dont rob it , just rob the previous one
    private int[] dp;
    
    // Time: O(n), Space: O(n)
    public int rob(int[] nums) {
        dp = new int[(nums.length <= 0) ? 1 : nums.length + 1];
        dp[0] = 0;
        if(nums.length == 0) return 0;
        
        dp[1] = nums[0];
        
        for(int i = 2; i <= nums.length; i++){
            dp[i] = Math.max(dp[i - 2] + nums[i - 1], dp[i - 1]);
        }
        
        return dp[nums.length];
    }
    
}